### PR TITLE
[VectorExt] Move VectorExt from llvm-external-projects to Codegen/Dialect

### PR DESCRIPTION
**Pull request by Groverkss**
This patch moves the VectorExt dialect from llvm-external-projects to compiler/Codegen/Dialect. Keeping the dialect externally with a different code style is a headache.

This patch also drops a redundant iterator test from VectorExt which is extensively tested by VectorDistribution now.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,172 @@
+# Copyright 2024 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+load("//build_tools/bazel:build_defs.oss.bzl", "iree_compiler_cc_library", "iree_gentbl_cc_library", "iree_tablegen_doc", "iree_td_library")
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/BUILD.bazel:7]: This file should be using the loaded `iree_tablegen_doc` function. That generates markdown files from tablegen files and we postprocess the generated files to host them on https://iree.dev/reference/mlir-dialects/.

I can help with that if you'd like. Some changes will need to be made across various files (searching for `IREEVectorExt` under `docs/` will show them)

Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/BUILD.bazel:7]: I'd like to do it. I'll send a follow-up patch after this patch (if that's okay with you). Want to keep this patch mostly as a move.

Comment (ScottTodd) [reply compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/BUILD.bazel:7]: I would prefer to include the changes here.

Bazel at least should complain about the unused load, which is why I anchored on that for this comment.

The website publication workflow runs automatically after each commit is merged to `main`: https://github.com/iree-org/iree/actions/workflows/publish_website.yml. That does not run any tests on presubmit and I'm not sure how that will handle these files moving without the scripts being updated.

Here's a model PR showing where to make edits: https://github.com/iree-org/iree/pull/17743. Docs for previewing website changes locally: https://github.com/iree-org/iree/tree/main/docs/website#how-to-edit-this-documentation

Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/BUILD.bazel:7]: Ok, let me try to include them

**Review by MaheshRavishankar — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/BUILD.bazel (hunk 1 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/BUILD.bazel:7]: I added doc gen and checked that it produces the docs properly. Can you have a look too if I did right?

File: docs/website/mkdocs.yml [hunk 2]
```diff
@@ -168,6 +168,7 @@ nav:
               - IO/Parameters: "reference/mlir-dialects/IOParameters.md"
               - IREECodegen: "reference/mlir-dialects/IREECodegen.md"
               - IREEGPU: "reference/mlir-dialects/IREEGPU.md"
+              - IREEVectorExt: "reference/mlir-dialects/IREEVectorExt.md"
```
Comment (ScottTodd) [inline docs/website/mkdocs.yml:171]: Also remove line 180
```diff
          # Dialects from llvm-external-projects/iree-dialects/
          - "Public dialects":
              - IREEInput: "reference/mlir-dialects/IREEInput.md"
-             - IREEVectorExt: "reference/mlir-dialects/IREEVectorExt.md"
```

**Review by ScottTodd — APPROVED** [review]
Thanks!

File: docs/website/mkdocs.yml (hunk 2 above)
Comment (Groverkss) [reply docs/website/mkdocs.yml:171]: fixed, thanks

**Comment by hanhanW** [comment]
Awesome, thanks for doing it! I've been wanting this after I moved LinalgExt from llvm-external-projects to iree main source. I did not touch VectorExt because people were actively developing the dialect.

**Review by kuhar — APPROVED** [review]

**Comment by ScottTodd** [comment]
Docs are live, the changes there worked as expected: https://iree.dev/reference/mlir-dialects/IREEVectorExt/

**Comment by ScottTodd** [comment]
The nightly macOS x64 compiler build failed after this:
https://github.com/iree-org/iree/actions/runs/10213277914/job/28258334819#step:8:5486

```
[4768/8277] Building CXX object compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/CMakeFiles/iree_compiler_Codegen_Dialect_VectorExt_IR_IREEVectorExtDialect.objects.dir/VectorExtAttrs.cpp.o
FAILED: compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/CMakeFiles/iree_compiler_Codegen_Dialect_VectorExt_IR_IREEVectorExtDialect.objects.dir/VectorExtAttrs.cpp.o 
/usr/local/bin/ccache /Applications/Xcode_15.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -I/Users/runner/work/iree/iree -I/Users/runner/work/iree/iree/build-macos -I/Users/runner/work/iree/iree/third_party/llvm-project/llvm/include -I/Users/runner/work/iree/iree/build-macos/llvm-project/include -I/Users/runner/work/iree/iree/third_party/llvm-project/mlir/include -I/Users/runner/work/iree/iree/build-macos/llvm-project/tools/mlir/include -I/Users/runner/work/iree/iree/third_party/llvm-project/lld/include -I/Users/runner/work/iree/iree/build-macos/llvm-project/tools/lld/include -I/Users/runner/work/iree/iree/compiler/src -I/Users/runner/work/iree/iree/build-macos/compiler/src -O2 -g  -std=gnu++17 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk -mmacosx-version-min=13.6 -fPIC -fvisibility=hidden -fno-rtti -fno-exceptions -Werror -Wall -Wno-error=deprecated-declarations -Wno-ambiguous-member-template -Wno-char-subscripts -Wno-extern-c-compat -Wno-gnu-alignof-expression -Wno-gnu-variable-sized-type-not-at-end -Wno-ignored-optimization-argument -Wno-invalid-offsetof -Wno-invalid-source-encoding -Wno-mismatched-tags -Wno-pointer-sign -Wno-reserved-user-defined-literal -Wno-return-type-c-linkage -Wno-self-assign-overloaded -Wno-sign-compare -Wno-signed-unsigned-wchar -Wno-strict-overflow -Wno-trigraphs -Wno-unknown-pragmas -Wno-unknown-warning-option -Wno-unused-command-line-argument -Wno-unused-const-variable -Wno-unused-function -Wno-unused-local-typedef -Wno-unused-private-field -Wno-user-defined-warnings -Wno-missing-braces -Wctad-maybe-unsupported -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wformat-security -Wgnu-redeclared-enum -Wimplicit-fallthrough -Winfinite-recursion -Wliteral-conversion -Wnon-virtual-dtor -Woverloaded-virtual -Wpointer-arith -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wthread-safety -Wthread-safety-beta -Wunused-comparison -Wvla -fno-lax-vector-conversions -fno-omit-frame-pointer -fmacro-prefix-map=/Users/runner/work/iree/iree=iree -MD -MT compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/CMakeFiles/iree_compiler_Codegen_Dialect_VectorExt_IR_IREEVectorExtDialect.objects.dir/VectorExtAttrs.cpp.o -MF compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/CMakeFiles/iree_compiler_Codegen_Dialect_VectorExt_IR_IREEVectorExtDialect.objects.dir/VectorExtAttrs.cpp.o.d -o compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/CMakeFiles/iree_compiler_Codegen_Dialect_VectorExt_IR_IREEVectorExtDialect.objects.dir/VectorExtAttrs.cpp.o -c /Users/runner/work/iree/iree/compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/VectorExtAttrs.cpp
/Users/runner/work/iree/iree/compiler/src/iree/compiler/Codegen/Dialect/VectorExt/IR/VectorExtAttrs.cpp:221:12: error: implicit conversion turns string literal into bool: 'const char[44]' to 'bool' [-Werror,-Wstring-conversion]
    assert("Invalid dimension! Expected lane dimension.");
    ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/assert.h:99:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __ASSERT_FILE_NAME, __LINE__, #e) : (void)0)
                      ~ ^
1 error generated.
```

Looks like the style we use for this sort of `assert("string")` is `assert(false && "string");`

**Comment by ScottTodd** [comment]
Pushed a fix for the macos compiler build: https://github.com/iree-org/iree/commit/74790bda7fa86bec9bfc0e4112f4f8e1fff92fb7

**Pull request merged**